src/modules/src/position_controller_indi.c

```diff
@@ -0,0 +1,354 @@
+}
+
+// Thrust filter
+static inline void filter_thrust(Butterworth2LowPass *filter, float *old_thrust, float *new_thrust) 
+{
+	*new_thrust = update_butterworth_2_low_pass(&filter[0], *old_thrust);
+}
+
+
+// Computes transformation matrix from body frame (index B) into NED frame (index O)
+void m_ob(struct Angles att, float matrix[3][3]) {
+
+	matrix[0][0] = cosf(att.theta)*cosf(att.psi);
+	matrix[0][1] = sinf(att.phi)*sinf(att.theta)*cosf(att.psi) - cosf(att.phi)*sinf(att.psi); 
+	matrix[0][2] = cosf(att.phi)*sinf(att.theta)*cosf(att.psi) + sinf(att.phi)*sinf(att.psi);
+	matrix[1][0] = cosf(att.theta)*sinf(att.psi);
+	matrix[1][1] = sinf(att.phi)*sinf(att.theta)*sinf(att.psi) + cosf(att.phi)*cosf(att.psi);
+	matrix[1][2] = cosf(att.phi)*sinf(att.theta)*sinf(att.psi) - sinf(att.phi)*cosf(att.psi);
+	matrix[2][0] = -sinf(att.theta);
+	matrix[2][1] = sinf(att.phi)*cosf(att.theta);
+	matrix[2][2] = cosf(att.phi)*cosf(att.theta);
+}
+
+
+void positionControllerINDIInit(void)
+{
+	// Re-initialize filters
+	position_indi_init_filters();
+}
+
+
+void positionControllerINDI(const sensorData_t *sensors, 
+	 						const state_t *state,
+	 						vector_t *refOuterINDI) { 
+
+	// Read states (position, velocity)
+	posS_x = state->position.x;
+	posS_y = -state->position.y;
+	posS_z = -state->position.z;
+	velS_x = state->velocity.x;
+	velS_y = -state->velocity.y;
+	velS_z = -state->velocity.z;
+	gyr_p = sensors->gyro.x;
+	gyr_q = sensors->gyro.y;
+	gyr_r = sensors->gyro.z; 
+
+	// Position controller (Proportional)
+	positionRef.x = pos_set_x;
+	positionRef.y = pos_set_y;
+	positionRef.z = pos_set_z;
+	velocityRef.x = K_xi_x*(positionRef.x - posS_x);
+	velocityRef.y = K_xi_y*(positionRef.y - posS_y);
+	velocityRef.z = K_xi_z*(positionRef.z - posS_z);
+
+	// Velocity controller (Proportional)
+	indiOuter.linear_accel_ref.x = K_dxi_x*(velocityRef.x - velS_x);
+	indiOuter.linear_accel_ref.y = K_dxi_y*(velocityRef.y - velS_y);
+	indiOuter.linear_accel_ref.z = K_dxi_z*(velocityRef.z - velS_z); 
+
+	// Acceleration controller (INDI)
+	// Read lin. acceleration (Body-fixed) obtained from sensors CHECKED
+	indiOuter.linear_accel_s.x = (sensors->acc.x)*9.81f;
+	indiOuter.linear_accel_s.y = (-sensors->acc.y)*9.81f;
+	indiOuter.linear_accel_s.z = (-sensors->acc.z)*9.81f;
+
+	// Filter lin. acceleration 
+	filter_ddxi(indiOuter.ddxi, &indiOuter.linear_accel_s, &indiOuter.linear_accel_f);
+
+	// Obtain actual attitude values (in deg)
+	indiOuter.attitude_s.phi = state->attitude.roll; 
+	indiOuter.attitude_s.theta = state->attitude.pitch;
+	indiOuter.attitude_s.psi = -state->attitude.yaw;
+	filter_ang(indiOuter.ang, &indiOuter.attitude_s, &indiOuter.attitude_f);
+
+
+	// Actual attitude (in rad)
+	struct Angles att = {
+		.phi = indiOuter.attitude_f.phi/180*PI,
+		.theta = indiOuter.attitude_f.theta/180*PI,
+		.psi = indiOuter.attitude_f.psi/180*PI,
+	};
+
+	// Compute transformation matrix from body frame (index B) into NED frame (index O)
+	float M_OB[3][3] = {0};
```

Comment:
In general I would prefer if matrix acceleration of the stm is used, so using the arm_matrix_instance_f32 and such like in kalman_core.c. But this is not a huge part of the code. Let us know if you want to change this or rather save this for another pull request

Reply:
If possible, I would leave it for the next pull request.

Reply:
Oke no problem